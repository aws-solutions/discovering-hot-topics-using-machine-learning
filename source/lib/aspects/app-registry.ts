#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the 'License'). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as applicationinsights from '@aws-cdk/aws-applicationinsights';
import * as appreg from '@aws-cdk/aws-servicecatalogappregistry';
import * as crypto from 'crypto';

export class AppRegistry extends cdk.Construct implements cdk.IAspect {
    /**
     * Name of the solution as set through from cdk.json
     */
    private solutionName: string;

    /**
     * Name of the application used to create an entry in AppRegistry as set through cdk.json
     */
    private applicationName: string;

    /**
     * Solution ID as set through cdk.json
     */
    private solutionID: string;

    /**
     * Solution version as set through cdk.json
     */
    private solutionVersion: string;

    /**
     * An AttributeGroupName initialized in the constructor of this class
     */
    private attributeGroupName: string;

    /**
     * An application type attribute initialized in the constructor of this class
     */
    private applicationType: string;

    /**
     * The instance of application that the solution stacks should be associated with
     */
    private application: appreg.Application;

    constructor(scope: cdk.Construct, id: string) {
        super(scope, id);
        this.solutionName = scope.node.tryGetContext('solution_name');
        this.applicationName = scope.node.tryGetContext('app_registry_name');
        this.solutionID = scope.node.tryGetContext('solution_id');
        this.solutionVersion = scope.node.tryGetContext('solution_version');
        this.applicationType = scope.node.tryGetContext('application_type');
    }

    /**
     * Method invoked as a `Visitor` pattern to inject aspects during cdk synthesis
     *
     * @param node
     */
    public visit(node: cdk.IConstruct): void {
        if (node instanceof cdk.Stack) {
            if (!node.nested) {
                // parent stack
                const stack = node as cdk.Stack;
                this.createAppForAppRegistry();
                this.application.associateStack(stack);
                this.createAttributeGroup();
                this.addTagsforApplication();
            }
        }

        if (node instanceof cdk.NestedStack) {
            if (!this.application) {
                this.createAppForAppRegistry();
            }

            const stack = node as cdk.NestedStack;

            const nestedStackResourceAssociation = new appreg.CfnResourceAssociation(
                this,
                `ResourceAssociation${crypto.randomUUID()}`,
                {
                    application: this.application.applicationId,
                    resource: node.stackId,
                    resourceType: 'CFN_STACK'
                }
            );

            const cfnNestedStackResource = stack.nestedStackResource as cdk.CfnResource;
            if (cfnNestedStackResource.hasOwnProperty('cfnOptions')) {
                nestedStackResourceAssociation.cfnOptions.condition = cfnNestedStackResource.cfnOptions.condition;
            }
        }
    }

    /**
     * Method to initialize an Application in AppRegistry service
     *
     * @returns - Instance of AppRegistry's Application class
     */
    private createAppForAppRegistry(): void {
        this.application = new appreg.Application(this, 'RegistrySetup', {
            applicationName: cdk.Fn.join('-', ['App', cdk.Aws.STACK_NAME, this.applicationName]),
            description: `Service Catalog application to track and manage all your resources for the solution ${this.solutionName}`
        });
    }

    /**
     * Method to add tags to the AppRegistry's Application instance
     *
     */
    private addTagsforApplication(): void {
        if (!this.application) {
            this.createAppForAppRegistry();
        }
        cdk.Tags.of(this.application).add('Solutions:SolutionID', this.solutionID);
        cdk.Tags.of(this.application).add('Solutions:SolutionName', this.solutionName);
        cdk.Tags.of(this.application).add('Solutions:SolutionVersion', this.solutionVersion);
        cdk.Tags.of(this.application).add('Solutions:ApplicationType', this.applicationType);
    }

    /**
     * Method to create AttributeGroup to be associated with the Application's instance in AppRegistry
     *
     */
    private createAttributeGroup(): void {
        if (!this.application) {
            this.createAppForAppRegistry();
        }
        this.application.associateAttributeGroup(
            new appreg.AttributeGroup(this, 'AppAttributes', {
                attributeGroupName: `AttrGrp-${cdk.Aws.STACK_NAME}`,
                description: 'Attributes for Solutions Metadata',
                attributes: {
                    applicationType: this.applicationType,
                    version: this.solutionVersion,
                    solutionID: this.solutionID,
                    solutionName: this.solutionName
                }
            })
        );
    }
}

