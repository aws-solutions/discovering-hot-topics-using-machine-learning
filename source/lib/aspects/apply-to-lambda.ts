#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as lambda from '@aws-cdk/aws-lambda';
import * as lambda_python from '@aws-cdk/aws-lambda-python';
import * as cdk from '@aws-cdk/core';
import { NodejsLayerVersion } from '../awsnodejs-lambda-layer/layer';

export class ApplytoLambda extends cdk.Construct implements cdk.IAspect {
    readonly pythonLayer: lambda.LayerVersion;
    readonly nodejsLayer: lambda.LayerVersion;

    constructor(scope: cdk.Construct, id: string) {
        super(scope, id);

        this.pythonLayer = new lambda_python.PythonLayerVersion(this, 'PythonLibLayer', {
            entry: 'lambda/layers/python_lambda_layer',
            description: 'This layer has boto config initialization and logging functions',
            compatibleRuntimes: [lambda.Runtime.PYTHON_3_8, lambda.Runtime.PYTHON_3_9]
        });

        this.nodejsLayer = new NodejsLayerVersion(this, id, {
            entry: 'lambda/layers/aws-nodesdk-custom-config',
            compatibleRuntimes: [lambda.Runtime.NODEJS_14_X],
            description: 'This layer configures AWS Node SDK initialization'
        });
    }

    public visit(node: cdk.IConstruct): void {
        // the various rules that apply to the lambda function
        this.applyUserAgentAspect(node); // add user agent for node and python
    }

    private applyUserAgentAspect(node: cdk.IConstruct): void {
        const solutionID = node.node.tryGetContext('solution_id');
        const solutionVersion = node.node.tryGetContext('solution_version');

        if (node instanceof lambda.Function) {
            if (node.runtime.family == lambda.RuntimeFamily.NODEJS) {
                node.addLayers(this.nodejsLayer);
                node.addEnvironment(
                    'AWS_SDK_USER_AGENT',
                    `{ "customUserAgent": "AwsSolution/${solutionID}/${solutionVersion}" }`
                );
            } else if (node.runtime.family == lambda.RuntimeFamily.PYTHON) {
                node.addLayers(this.pythonLayer);
                node.addEnvironment(
                    'AWS_SDK_USER_AGENT',
                    `{ "user_agent_extra": "AwsSolution/${solutionID}/${solutionVersion}" }`
                );
            }
        }
    }
}
