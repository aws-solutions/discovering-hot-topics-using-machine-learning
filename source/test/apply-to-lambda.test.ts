/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/

import * as cdk from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { ApplytoLambda } from '../lib/aspects/apply-to-lambda';

test('test adding custom config', () => {
    const stack = new cdk.Stack();
    new ApplytoLambda(stack, 'TestConfig');
    Template.fromStack(stack).hasResourceProperties('AWS::Lambda::LayerVersion', {
        CompatibleRuntimes: ['python3.11'],
        Content: {}
    });
});

test('visting node lambda runtimes and adding aspects', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app);
    new lambda.Function(stack, 'testFunction', {
        code: lambda.Code.fromAsset(`${__dirname}/../lambda/ingestion-producer`),
        runtime: lambda.Runtime.NODEJS_20_X,
        handler: 'index.handler'
    });

    cdk.Aspects.of(stack).add(new ApplytoLambda(stack, 'testConfigWithNode'));
    Template.fromStack(stack).hasResourceProperties('AWS::Lambda::Function', {
        Runtime: 'nodejs20.x',
        Environment: {
            Variables: {
                'AWS_SDK_USER_AGENT': '{ "customUserAgent": "AwsSolution/undefined/undefined" }'
            }
        },
        Handler: 'index.handler',
        Role: {},
        Code: {
            S3Bucket: {},
            S3Key: Match.anyValue()
        }
    });

    Template.fromStack(stack).hasResource('AWS::Lambda::LayerVersion', {
        'Type': 'AWS::Lambda::LayerVersion'
    });
});

test('visting node python runtimes and adding aspects', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app);
    new lambda.Function(stack, 'testFunction', {
        code: lambda.Code.fromAsset(`${__dirname}/../lambda/quicksight-custom-resources`),
        runtime: lambda.Runtime.PYTHON_3_11,
        handler: 'handler'
    });

    cdk.Aspects.of(stack).add(new ApplytoLambda(stack, 'testConfigWithNode'));
    Template.fromStack(stack).hasResource('AWS::Lambda::Function', {
        Type: 'AWS::Lambda::Function',
        Properties: {
            Runtime: 'python3.11',
            Environment: {
                Variables: {
                    'AWS_SDK_USER_AGENT': '{ "user_agent_extra": "AwsSolution/undefined/undefined" }'
                }
            },
            Handler: 'handler',
            Layers: [{}],
            Role: {},
            Code: {
                S3Bucket: {},
                S3Key: Match.anyValue()
            }
        }
    });

    Template.fromStack(stack).hasResource('AWS::Lambda::LayerVersion', {
        'Type': 'AWS::Lambda::LayerVersion'
    });
});
