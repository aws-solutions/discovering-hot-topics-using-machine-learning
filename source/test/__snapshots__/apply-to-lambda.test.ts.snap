// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test adding custom config 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeArtifactHash7E1A96FC": Object {
      "Description": "Artifact hash for asset \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3Bucket0BACA0F3": Object {
      "Description": "S3 bucket for asset \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3": Object {
      "Description": "S3 key for asset version \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3ArtifactHash3A828C70": Object {
      "Description": "Artifact hash for asset \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3Bucket6ADCD2CE": Object {
      "Description": "S3 bucket for asset \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6": Object {
      "Description": "S3 key for asset version \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TestConfigDFE7B72C": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "nodejs14.x",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3Bucket0BACA0F3",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This layer configures AWS Node SDK initialization",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "TestConfigPythonLibLayer672F8799": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "python3.8",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3Bucket6ADCD2CE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This layer has boto config initialization and logging functions",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
  },
}
`;

exports[`visting node lambda runtimes and adding aspects 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652aArtifactHash982626B0": Object {
      "Description": "Artifact hash for asset \\"562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652a\\"",
      "Type": "String",
    },
    "AssetParameters562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652aS3Bucket5025E55C": Object {
      "Description": "S3 bucket for asset \\"562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652a\\"",
      "Type": "String",
    },
    "AssetParameters562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652aS3VersionKey614E5FAD": Object {
      "Description": "S3 key for asset version \\"562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652a\\"",
      "Type": "String",
    },
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeArtifactHash7E1A96FC": Object {
      "Description": "Artifact hash for asset \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3Bucket0BACA0F3": Object {
      "Description": "S3 bucket for asset \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3": Object {
      "Description": "S3 key for asset version \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3ArtifactHash3A828C70": Object {
      "Description": "Artifact hash for asset \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3Bucket6ADCD2CE": Object {
      "Description": "S3 bucket for asset \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6": Object {
      "Description": "S3 key for asset version \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "testConfigWithNodeDDA76C99": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "nodejs14.x",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3Bucket0BACA0F3",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This layer configures AWS Node SDK initialization",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "testConfigWithNodePythonLibLayer75C81C17": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "python3.8",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3Bucket6ADCD2CE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This layer has boto config initialization and logging functions",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "testFunction483F4CBE": Object {
      "DependsOn": Array [
        "testFunctionServiceRoleFEC29B6F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652aS3Bucket5025E55C",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652aS3VersionKey614E5FAD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters562b22572f5b1c96e642fd6530a7f212a75859278d0a3c37118347ab5e4f652aS3VersionKey614E5FAD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_SDK_USER_AGENT": "{ \\"customUserAgent\\": \\"AwsSolution/undefined/undefined\\" }",
          },
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "testConfigWithNodeDDA76C99",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "testFunctionServiceRoleFEC29B6F",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testFunctionServiceRoleFEC29B6F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`visting node python runtimes and adding aspects 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9ArtifactHashD6A510A4": Object {
      "Description": "Artifact hash for asset \\"993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9\\"",
      "Type": "String",
    },
    "AssetParameters993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9S3Bucket7D3853F1": Object {
      "Description": "S3 bucket for asset \\"993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9\\"",
      "Type": "String",
    },
    "AssetParameters993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9S3VersionKey8E9FB346": Object {
      "Description": "S3 key for asset version \\"993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9\\"",
      "Type": "String",
    },
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeArtifactHash7E1A96FC": Object {
      "Description": "Artifact hash for asset \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3Bucket0BACA0F3": Object {
      "Description": "S3 bucket for asset \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3": Object {
      "Description": "S3 key for asset version \\"a7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeee\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3ArtifactHash3A828C70": Object {
      "Description": "Artifact hash for asset \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3Bucket6ADCD2CE": Object {
      "Description": "S3 bucket for asset \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
    "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6": Object {
      "Description": "S3 key for asset version \\"d8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "testConfigWithNodeDDA76C99": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "nodejs14.x",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3Bucket0BACA0F3",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa7fd33e471aee73db5f46a802639572a716ebc4691e30f037db9615e5422eeeeS3VersionKey88BCADE3",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This layer configures AWS Node SDK initialization",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "testConfigWithNodePythonLibLayer75C81C17": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "python3.8",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3Bucket6ADCD2CE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd8876cc7c42a8c1f1336baad719c770a079eefa0a234b4540ffa46a70b397bf3S3VersionKeyBFE2A5C6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This layer has boto config initialization and logging functions",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "testFunction483F4CBE": Object {
      "DependsOn": Array [
        "testFunctionServiceRoleFEC29B6F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9S3Bucket7D3853F1",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9S3VersionKey8E9FB346",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters993a3efadb0d38ccd49e129ec9c953993426e0f257bcb696bf7ba410fe4034f9S3VersionKey8E9FB346",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_SDK_USER_AGENT": "{ \\"user_agent_extra\\": \\"AwsSolution/undefined/undefined\\" }",
          },
        },
        "Handler": "handler",
        "Layers": Array [
          Object {
            "Ref": "testConfigWithNodePythonLibLayer75C81C17",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "testFunctionServiceRoleFEC29B6F",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testFunctionServiceRoleFEC29B6F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
